### PR TITLE
[ClusterObjects] Add kIsFabricScoped to all cluster objects

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem

Should make fabric scoped stuff safer

#### Change overview
- Add kIsFabricScoped to all cluster objects
- Check kIsFabricScoped in ListEncodeHelper::Encode
- Add IsBasicType for scalar types.

#### Testing
Unit tests should covered these code.

### CODE REVIEW

**Review by bzbarsky-apple — APPROVED** [review]
Thank you for doing this!

**Review by andy31415 — APPROVED** [review]

File: src/app/data-model/FabricScoped.h [hunk 1]
```diff
@@ -18,31 +18,39 @@
...
 namespace app {
 namespace DataModel {
 
-/*
- * Check whether a cluster object struct is fabric scoped.
- *   A fabric scoped struct contains a field of "FabricIndex" type, however, we cannot tell the difference between that field and
- * other uint8_t fields. Thus we add a GetFabricIndex member function for getting the fabric id. Here, IsFabricScoped checks the
- * presence of GetFabricIndex function. This template can be used with std::enable_if.
+/**
+ * IsFabricScoped checks whether the given type is fabric scoped, non-basic types (i.e. cluster objects), should provide a
```
Comment (bzbarsky-apple) [inline src/app/data-model/FabricScoped.h:30]:
```suggestion
 * IsFabricScoped checks whether the given type is fabric scoped.  Non-basic types (i.e. cluster objects), should provide a
```

**Comment by woody-apple** [comment]
Removing fast track.

**Review by woody-apple — APPROVED** [review]

**Comment by erjiaqing** [comment]
Pushed change after Yunhan's PR for fabric scoped event encode.

**Review by saurabhst — APPROVED** [review]

**Pull request merged**